File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles, h):
        left = 1
        right = max(piles)

        while left <= right:
            k = (left + right) // 2

            hours = 0

            for pile in piles:
                hours += (pile + k - 1) // k

            if hours <= h:
                right = k - 1
            else:
                left = k + 1

        return left

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
class Solution {
    int maxDiameter;
    public int diameterOfBinaryTree(TreeNode root) {
        maxDiameter = 0;
        height(root);
        return maxDiameter;
    }
    
    public int height(TreeNode root) {
        if(root == null)
            return -1;
        int left = height(root.left);
        int right = height(root.right);
        int h = 1 + Math.max(left, right);
        maxDiameter = Math.max(maxDiameter, left + right + 2);
        return h;
    }
}